File: internal/modules/semester/domain/semester.go
package domain

import (
	"context"
	"time"

	mkDomain "siakad-pro/internal/modules/matakuliah/domain"
)

type Semester struct {
	ID         string                `json:"id" gorm:"primaryKey;type:varchar(255)"`
	Nomor      int                   `json:"nomor" gorm:"uniqueIndex;not null"`
	MinSKS     int                   `json:"min_sks" gorm:"not null;default:18"`
	MaxSKS     int                   `json:"max_sks" gorm:"not null;default:24"`
	MataKuliah []*SemesterMataKuliah `json:"mata_kuliah,omitempty" gorm:"foreignKey:SemesterID"`
	CreatedAt  time.Time             `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt  time.Time             `json:"updated_at" gorm:"autoUpdateTime"`
}

type SemesterMataKuliah struct {
	ID           string               `json:"id" gorm:"primaryKey;type:varchar(255)"`
	SemesterID   string               `json:"semester_id" gorm:"type:varchar(255);uniqueIndex:idx_sem_mk;not null"`
	Semester     *Semester            `json:"semester,omitempty" gorm:"foreignKey:SemesterID"`
	MataKuliahID string               `json:"mata_kuliah_id" gorm:"type:varchar(255);uniqueIndex:idx_sem_mk;not null"`
	MataKuliah   *mkDomain.MataKuliah `json:"mata_kuliah,omitempty" gorm:"foreignKey:MataKuliahID"`
	Kategori     string               `json:"kategori" gorm:"type:varchar(50);not null;default:'wajib'"`
	CreatedAt    time.Time            `json:"created_at" gorm:"autoCreateTime"`
}

const (
	KategoriWajib   = "wajib"
	KategoriPilihan = "pilihan"

	MaxPertemuan = 16
	MinSemester  = 1
	MaxSemester  = 8
)

type CreateSemesterRequest struct {
	Nomor  int `json:"nomor" validate:"required,min=1,max=8"`
	MinSKS int `json:"min_sks" validate:"required,min=1"`
	MaxSKS int `json:"max_sks" validate:"required,min=1"`
}

type UpdateSemesterRequest struct {
	MinSKS *int `json:"min_sks"`
	MaxSKS *int `json:"max_sks"`
}

type AssignMataKuliahRequest struct {
	MataKuliahID string `json:"mata_kuliah_id" validate:"required"`
	Kategori     string `json:"kategori" validate:"required"`
}


type SemesterRepository interface {
	Create(ctx context.Context, s *Semester) error
	GetAll(ctx context.Context) ([]*Semester, error)
	GetByID(ctx context.Context, id string) (*Semester, error)
	GetByNomor(ctx context.Context, nomor int) (*Semester, error)
	Update(ctx context.Context, s *Semester) error
	Delete(ctx context.Context, id string) error

	AssignMataKuliah(ctx context.Context, sm *SemesterMataKuliah) error
	UnassignMataKuliah(ctx context.Context, semesterID string, mkID string) error
	GetSemesterMataKuliah(ctx context.Context, semesterID string) ([]*SemesterMataKuliah, error)
	GetTotalSKS(ctx context.Context, semesterID string) (int, error)


}

type SemesterService interface {
	Create(ctx context.Context, req CreateSemesterRequest) (*Semester, error)
	GetAll(ctx context.Context) ([]*Semester, error)
	GetByID(ctx context.Context, id string) (*Semester, error)
	Update(ctx context.Context, id string, req UpdateSemesterRequest) (*Semester, error)
	Delete(ctx context.Context, id string) error

	AssignMataKuliah(ctx context.Context, semesterID string, req AssignMataKuliahRequest) (*SemesterMataKuliah, error)
	UnassignMataKuliah(ctx context.Context, semesterID string, mkID string) error


}

File: frontend/src/lib/services/semester.ts
import { fetchApi } from './api';
import type { ApiResponse, Semester, SemesterMataKuliah, Pertemuan } from '../types';

export const semesterService = {
	getAll: async () => {
		return fetchApi<ApiResponse<Semester[]>>('/semester');
	},
	getById: async (id: string) => {
		return fetchApi<ApiResponse<Semester>>(`/semester/${id}`);
	},
	create: async (data: { nomor: number; min_sks: number; max_sks: number }) => {
		return fetchApi<ApiResponse<Semester>>('/semester', {
			method: 'POST',
			body: JSON.stringify(data)
		});
	},
	update: async (id: string, data: { min_sks?: number; max_sks?: number }) => {
		return fetchApi<ApiResponse<Semester>>(`/semester/${id}`, {
			method: 'PUT',
			body: JSON.stringify(data)
		});
	},
	delete: async (id: string) => {
		return fetchApi<ApiResponse<null>>(`/semester/${id}`, {
			method: 'DELETE'
		});
	},
	activate: async (id: string) => {
		return fetchApi<ApiResponse<null>>(`/semester/${id}/activate`, {
			method: 'PUT'
		});
	},
	assignMataKuliah: async (id: string, mata_kuliah_id: string, kategori: string = 'wajib') => {
		return fetchApi<ApiResponse<SemesterMataKuliah>>(`/semester/${id}/matakuliah`, {
			method: 'POST',
			body: JSON.stringify({ mata_kuliah_id, kategori })
		});
	},
	unassignMataKuliah: async (id: string, mkId: string) => {
		return fetchApi<ApiResponse<null>>(`/semester/${id}/matakuliah/${mkId}`, {
			method: 'DELETE'
		});
	},
	getPertemuan: async (kelasId: string, semesterId: string) => {
		return fetchApi<ApiResponse<Pertemuan[]>>(`/semester/pertemuan?kelas_id=${kelasId}&semester_id=${semesterId}`);
	},
	catatPertemuan: async (kelasId: string, semesterId: string, topik: string) => {
		return fetchApi<ApiResponse<Pertemuan>>('/semester/pertemuan', {
			method: 'POST',
			body: JSON.stringify({ kelas_id: kelasId, semester_id: semesterId, topik })
		});
	},
	setSKSProdi: async (id: string, configs: { program_studi_id: string; min_sks: number; max_sks: number }[]) => {
		return fetchApi<ApiResponse<any>>(`/semester/${id}/sks-prodi`, {
			method: 'PUT',
			body: JSON.stringify({ configs })
		});
	}
};
